source/Halibut.Tests/ServiceModel/PendingRequestQueueFixture.cs

```diff
@@ -0,0 +1,128 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Halibut.Diagnostics;
+using Halibut.ServiceModel;
+using Halibut.Transport.Protocol;
+using NUnit.Framework;
+
+namespace Halibut.Tests.ServiceModel
+{
+    public class PendingRequestQueueFixture
+    {
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task CanCancelAPendingRequestBeforeItIsDequeued(bool async)
+        {
+            // Arrange
+            const string endpoint = "poll://endpoint001";
+            var request = new RequestMessage
+            {
+                Id = Guid.NewGuid().ToString(),
+                Destination = new ServiceEndPoint(new Uri(endpoint), "thumbprint")
+            };
+            var cancellationTokenSource = new CancellationTokenSource();
+
+            // Act
+            var (pendingRequestQueue, task) = async ? await StartQueueAndWaitAsATask(endpoint, request, cancellationTokenSource)
+                    : await StartQueueAndWaitAsyncAsATask(endpoint, request, cancellationTokenSource);
+
+            cancellationTokenSource.Cancel();
+
+            Exception actualException = null;
+            try
+            {
+                await task;
+            }
+            catch (Exception ex)
+            {
+                actualException = ex;
+            }
+
+            // Assert
+            actualException.Should().NotBeNull().And.BeOfType<OperationCanceledException>();
+            var next = await pendingRequestQueue.DequeueAsync();
```

Comment:
Does this take 30s to return a null value?

Reply:
No, when a pending request is cancelled, it is marked as complete by Halibut but not removed from the queue! When dequeue is called it removes it from the queue at that point :( That's current behaviour!!

Reply:
crazy